### PR TITLE
space-age: forces a second branch for each planet

### PR DESCRIPTION
**Pull request by michaelglass**
previously, space age could hard code the test results for each planet. With two examples per planet, it makes more sense to calculate the results

(started here: https://github.com/exercism/elm/pull/250)

### CODE REVIEW

**Comment by michaelglass** [comment]
@cmccandless ordered by planet

File: exercises/space-age/canonical-data.json [hunk 1]
```diff
@@ -3,22 +3,22 @@
   "version": "1.3.0",
   "cases": [
     {
-      "description": "age on Earth",
+      "description": "age on Mercury",
```
Comment (cmccandless) [inline exercises/space-age/canonical-data.json]: I like the new `x earth years in mercury years` naming scheme. Perhaps we should apply the same scheme to the existing tests?

File: exercises/space-age/canonical-data.json [hunk 2]
```diff
@@ -3,22 +3,22 @@
   "version": "1.3.0",
```
Comment (cmccandless) [inline exercises/space-age/canonical-data.json]: Tests are being added, renamed, or removed, so we'll need to bump the minor version number.

`"version": "1.4.0",`

Comment (michaelglass) [reply exercises/space-age/canonical-data.json]: bump it twice?

Comment (cmccandless) [reply exercises/space-age/canonical-data.json]: Ah, no, don't do that. I didn't realize I was only looking at the changes for a single commit, not the whole PR. You're good already.

File: exercises/space-age/canonical-data.json (hunk 1 above)
Comment (michaelglass) [reply exercises/space-age/canonical-data.json]: reworked everything as 2 and 20 earth years

**Comment by michaelglass** [comment]
@cmccandless addressed your two tick boxes

File: exercises/space-age/canonical-data.json [hunk 3]
```diff
@@ -66,13 +120,31 @@
...
+    },
+    {
+      "description": "2 Earth years in Neptune years",
+      "property": "age",
+      "input": {
+        "planet": "Neptune",
+        "seconds": 63115200
+      },
+      "expected": 0.01
+    },
```
Comment (cmccandless) [inline exercises/space-age/canonical-data.json]: Comma after last array item is invalid.

```suggestion
    }
```

Comment (michaelglass) [reply exercises/space-age/canonical-data.json]: I've never seen this, "commit suggestion" thing. Neat!

File: exercises/space-age/canonical-data.json [hunk 4]
```diff
@@ -21,7 +30,16 @@
...
+      "description": "2 Earth years in Mercury years",
+      "property": "age",
+      "input": {
+        "planet": "Mercury",
+        "seconds": 63115200
+      },
+      "expected": 8.3
+    },
+    {
+      "description": "6 Earth years  in Venus years",
```
Comment (petertseng) [inline exercises/space-age/canonical-data.json]: it would probably be good to remove the extra space between `years` and `in`, for the purpose of consistency (no other case has this extra space)

File: exercises/space-age/canonical-data.json [hunk 5]
```diff
@@ -12,7 +12,16 @@
...
     },
     {
-      "description": "age on Mercury",
+      "description": "2 Earth years in Earth years",
+      "property": "age",
+      "input": {
+        "planet": "Earth",
+        "seconds": 63115200
+      },
+      "expected": 2
```
Comment (ErikSchierboom) [inline exercises/space-age/canonical-data.json]:
```suggestion
      "expected": 2.0
```

This is just something to make some test generators a bit more happy, the ones that discern between ints and float.

**Comment by bitfield** [comment]
I've literally never encountered a solution which hard-coded the results for each planet. If I did, I'd approve it. The tests are the tests; if the code passes the tests, it is correct. (That's not to say it couldn't be _improved_; that's what mentoring is for.)

**Comment by NobbZ** [comment]
No, just passing the tests is not the barrier. The barrier is
understanding.

If the student just hard coded the return values for each testcase he
proved to not have understood the exercise. I'd try to explain it to them
then.

John Arundel <notifications@github.com> schrieb am Sa., 20. Apr. 2019,
13:13:

**Comment by bitfield** [comment]
> If the student just hard coded the return values for each testcase he proved to not have understood the exercise.

This is a fundamental philosophical disagreement about what tests are for! (Which can't be resolved in a GitHub thread. I note it merely for the record, and to make it clear that intelligent people can disagree honourably about this issue.)

**Comment by SleeplessByte** [comment]
True. But _because_ we have an agreement about not adding tests for specific implementation and we have an agreement about what exercism is and is for, in this case, I agree with @nobbz.

**Comment by michaelglass** [comment]
I disagree, in so far as hardcoding the result is the minimal solution, and I think it's often easy to argue that the minimal solution is best, e.g. https://github.com/exercism/problem-specifications/pull/1512#issuecomment-485094254.

For me there are two questions:
1. What is it that exercism wants to teach?
2. Is it desirable to require mentors to intervene?

My opinions:
1. Personally, I think a more thorough test suite _teaches_ better practice. I also think the minimal implementation is often considered best. E.g. I find [this argument](https://github.com/exercism/problem-specifications/pull/1512#issuecomment-485094254) compelling.
2. The feedback loop of a test suite is much faster than a mentor. If I have to wait hours or days for a response, I've already lost flow. If a test fails, I know immediately. If I can document or codify my suggestions (e.g. through the readme, the test suite) then I'm providing students with more immediate feedback.

**Comment by bitfield** [comment]
> we have an agreement about not adding tests for specific implementation

Then doesn't this PR fall foul of that agreement?

> we have an agreement about what exercism is and is for

I think this conversation demonstrates otherwise.

**Comment by SleeplessByte** [comment]
> I think this conversation demonstrates otherwise.

You can find the [goal of exercism here](https://github.com/exercism/docs/blob/master/about/goal-of-exercism.md) and see [this document](https://github.com/exercism/automated-mentoring-support/blob/f2c7c647d92782481b84c46363766754bf803824/docs/guidelines.md) about what the automated mentoring project will approve and disapprove.

These two documents together are the agreement I am referring to :) 

> Then doesn't this PR fall foul of that agreement?

I think that's definitely an argument to be made. I don't see the point of adding these tests other than in the 1% of cases save a mentor from saying "but now it doesn't work for other input". For that reason I have _not_ ✔️ approved these changes myself.

**Comment by michaelglass** [comment]
1. What is the _harm_ that this PR introduces?
- increased complexity?

2. How can I participate in a way that feels more light-weight.
- as a new member of the community I wish that
    - I had put in less work on this PR.
    - Others had put in less work on this PR. Three of this PR's six reviewers put in time editing something that will (likely) get thrown out.
    - I had clearer guidelines around how to propose a change like this. (It appears: a small issue describing the test case is the approach).
    - It was more clear to me who is responsible for making decisions.

Now, of course if this PR is going to make Exercism _worse_ than it's better that discarded then held dear because of the effort. But if it's discarded, I'd love to makee it harder to waste contributor's time.

To that end: I think issue & PR templates would be a nice next step.

**Comment by SleeplessByte** [comment]
@michaelglass I think it's still very valuable that you made this PR. How we do things and why we do things aren't clear for everyone (including myself -- I had to look up the written docs) and discussions like these force all of us to up our game and look at the guideliness written in the various repo's. I apologise you put in so much work; but don't think this isn't fruitful. 

> I had clearer guidelines around how to propose a change like this. (It appears: a small issue describing the test case is the approach).

Did [CONTRIBUTING.md](https://github.com/exercism/problem-specifications/blob/master/CONTRIBUTING.md#problem-metadata) not suffice in this case? It says:
```md
We try to keep the generic descriptions generic--we avoid implementation-specific examples, and 
try not to be too prescriptive about suggesting how a problem might be solved.
```

> It was more clear to me who is responsible for making decisions.

Everyone here basically has "equal rights" as far as I can tell, but Jeremy (`iHiD`) and Katrina (kytrinyx) always have final and ultimate say. 

In this repository @petertseng takes care of a _lot_ of the important steps; e.g. I see him make sure all implementing tracks have a `sync` issue opened and interesting discussion around the merit of certain changes. Furthermore, some of the exercises are designed by people here such as the `resistor-xxx` series by @ErikSchierboom and @F3PiX, so we tend to ping them when a change is made to those exercises.

That said, the general consensus here is in `problem-specifications` that all us `track-maintainers` chime in when we want to and we merge if we get at least three ✔️ without strong ❌ .

> To that end: I think issue & PR templates would be a nice next steps.

I think this is a _great_ idea!

**Comment by michaelglass** [comment]
for #1512, it's clear that I was pushing for a specific implementation strategy. For this case, I don't think I am (unless the specific implementation is, "actually do what the readme says").

I'll follow-up with some proposals for issue templates when I get the time. In the mean time, should I close this PR?

**Comment by kytrinyx** [comment]
A few thoughts:

It sounds like here the key problem is that there was no discussion up front about what constraints the exercise has.

I think that it would be valuable to have a _short_ template, preferably within HTML comments, so that if the template pieces don't get deleted, we don't have to always look at the text. E.g. something like this.








In this case, I actually think that adding a few tests to ensure that the implementation is not hard-coded seems perfectly reasonable. That is generally considered a reasonable practice in TDD ("triangulation"). I didn't do it originally when I added the exercise, because it didn't occur to me that people would hard-code the values.

If we do add tests, I think that we should be careful to keep the values within the 32-bit range.

**Comment by cmccandless** [comment]
> E.g. something like this. *[invisible comments below]*

```
<!-- NOTE: changing exercise specifications has a lot of complicated considerations. You may want to open an issue to discuss before opening a PR. -->

<!-- What problem did you observe? How do you propose to fix it? -->

<!-- Please look at earlier PRs and commits for this exercise. What relevant points have been discussed previously? -->
```

**Comment by kytrinyx** [comment]
d'oh. Thank you @cmccandless !

**Comment by SleeplessByte** [comment]
IMO, the only _correct_ way to ensure people don't hardcode the value is randomly _generate_ the input (and output). Adding a second set of values can still make some people hardcode the values, albeit probably less.

I'm assuming there was a reason the tests for `two-fer` are only <nothing>. `Alice` and `Bob`? We still see people hardcode these as well.

👍 * 💯 on the comment in the PR template!

**Comment by sshine** [comment]
Perhaps we can document this policy against additional hardcoded unit tests to avoid hardcoded solutions so that @michaelglass'es efforts were not in vain, and so that others will not repeat the process of wasting time on similar PRs.

**Comment by michaelglass** [comment]
The "only correct test are quickcheck" argument similar to "perfect is the enemy of good". Does the metadata support generative test?

The "What about the two-fer tests" is in the realm of [whataboutism](https://en.wikipedia.org/wiki/Whataboutism).

When is a test change to an existing test suite valid?

**Comment by SleeplessByte** [comment]
@michaelglass my comment really isn't whataboutism. It's the first exercise in _most_ tracks and I am assuming here that the inputs were purposefully **not** generated. 

If we now are trending towards "let's make it harder for people to hard-code", I was going to suggest we actually make this change in _all_ exercises we see people hardcode, including `two-fer`.

**Comment by ErikSchierboom** [comment]
> If we now are trending towards "let's make it harder for people to hard-code", I was going to suggest we actually make this change in all exercises we see people hardcode, including two-fer.

Personally, I see very, very little hardcoded exercises. I see so little of that in fact, that I'm against adding test cases for that specific purpose :)

**Comment by SleeplessByte** [comment]
It's less than 1 in each hundred (going of the data I got for writing the analyzer)

**Comment by SleeplessByte** [comment]
@michaelglass Do you think something like [this](https://github.com/exercism/problem-specifications/pull/1511#issuecomment-486400333) (my last comment) would have been useful in the docs?

**Comment by michaelglass** [comment]
I think `Contributing` is already a huge doc to wrap your mind around, but I think it'd be useful for a PR template.

But, I think I would have considered this PR falling under this case

> completeness, e.g. a test is missing which allows for an incorrect implementation (for example tests were not covering a leading zero, or odd numbers).

**Comment by SleeplessByte** [comment]
Yes, I understand what you're saying. Thank you for spending the time ❤️ . I don't think it was wasted effort but I am sorry this went the way it did.
